atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStore.java

```diff
@@ -0,0 +1,234 @@
+                    "The timestamp is unreadable, though the backup is also unreadable! Please contact support.");
+            byte[] backupValue = MoreObjects.firstNonNull(currentBound, PtBytes.toBytes(defaultValue));
+            ByteBuffer casQueryBuffer = CassandraTimestampUtils.constructCheckAndSetMultipleQuery(
+                    ImmutableMap.of(
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            Pair.create(currentBound, CassandraTimestampUtils.INVALIDATED_VALUE.toByteArray()),
+                            CassandraTimestampUtils.BACKUP_COLUMN_NAME,
+                            Pair.create(currentBackupBound, backupValue)));
+            executeQueryUnchecked(client, casQueryBuffer);
+            return PtBytes.toLong(backupValue);
+        });
+    }
+
+    /**
+     * Restores a backup of an existing timestamp, if possible. Note that this method will throw if the timestamp
+     * backup is unreadable, *and* the current bound is also unreadable.
+     */
+    public synchronized void restoreFromBackup() {
+        clientPool().runWithRetry(client -> {
+            BoundData boundData = getCurrentBoundData(client);
+            byte[] currentBound = boundData.bound();
+            byte[] currentBackupBound = boundData.backupBound();
+
+            if (CassandraTimestampUtils.isValidTimestampData(currentBound)) {
+                Preconditions.checkState(currentBackupBound == null
+                                || !CassandraTimestampUtils.isValidTimestampData(currentBackupBound),
+                        "We had both backup and active bounds readable! This is unexpected; please contact support.");
+                log.info("[RESTORE] Didn't restore from backup, because the current timestamp is already readable.");
+                return null;
+            }
+
+            Preconditions.checkState(CassandraTimestampUtils.isValidTimestampData(currentBackupBound),
+                    "The timestamp backup is unreadable, so we cannot restore our backup. Please contact support.");
+            ByteBuffer casQueryBuffer = CassandraTimestampUtils.constructCheckAndSetMultipleQuery(
+                    ImmutableMap.of(
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            Pair.create(CassandraTimestampUtils.INVALIDATED_VALUE.toByteArray(), currentBackupBound),
+                            CassandraTimestampUtils.BACKUP_COLUMN_NAME,
+                            Pair.create(currentBackupBound, PtBytes.EMPTY_BYTE_ARRAY)));
+            executeQueryUnchecked(client, casQueryBuffer);
+            return null;
+        });
+    }
+
+    private void checkOperationWasApplied(Optional<Long> expected, long target, CqlResult result) {
+        if (!CassandraTimestampUtils.wasOperationApplied(result)) {
+            throw constructConcurrentTimestampStoreException(expected, target, result);
```

Comment:
Okay, made some changes to this. The store method now returns one of these:

```java
    @Value.Immutable
    interface StoreTimestampResult {
        boolean successful();
        Optional<Long> expected();
        Optional<Long> actual();
    }
```

All the exception handling logic is now in CTBS, which checks that the result is successful, or otherwise grabs the expected and actual values and uses them to build up the MRTSE.